### PR TITLE
Harden VertxRequestWriteSubsriber

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Fixes: #43399

Hardens the logic in `VertxRequestWriteSubsriber` to make is less brittle if the `Promise` being used in completed or failed externally to the class.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 1]
```diff
@@ -134,7 +139,17 @@ private void onErrorInternal(Throwable throwable) {
...
-        promise.fail(throwable);
+        if (!promise.tryFail(throwable)) {
+            // Seems the promise has already completed in some form.
+            // Attempt to associate this error with the existing failure.
+            Throwable cause = promise.future().cause();
+            if (cause != null) {
+                cause.addSuppressed(throwable);
+            } else {
+                // Turns out the future was completed as successfully externally, drop the error.
+                Operators.onErrorDropped(throwable, Context.of(contextView));
```
Comment (srnagar) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:150]: Should we log a debug message here?

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 2]
```diff
@@ -134,7 +139,17 @@ private void onErrorInternal(Throwable throwable) {
 
     private void resetRequest(Throwable throwable) {
         subscription.cancel();
-        promise.fail(throwable);
+        if (!promise.tryFail(throwable)) {
+            // Seems the promise has already completed in some form.
+            // Attempt to associate this error with the existing failure.
+            Throwable cause = promise.future().cause();
+            if (cause != null) {
+                cause.addSuppressed(throwable);
```
Comment (srnagar) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:147]: Log a debug message here so we don't lose the exception.

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        assertSame(reactiveError, writeSuppressed[0], "The reactive error should be the suppressed exception.");
+    }
+
+    private static void waitForCompletion(Promise<?> promise) throws InterruptedException {
+        Thread.sleep(1000);
+        for (int i = 0; i < 4; i++) {
+            if (promise.future().isComplete()) {
+                break;
+            }
+
```
Comment (anuchandy) [inline sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java:245]: Confirming, do we meant to put the thread.sleep inside loop?

**Review by anuchandy — APPROVED** [review]

File: sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java:245]: Haha, good catch!

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:147]: This should be logged by the `HttpLoggingPolicy`.

File: sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+        data.subscribe(subscriber);
+
+        waitForCompletion(promise);
+        assertTrue(promise.future().failed(), "The request should have failed.");
+        assertSame(writeError, promise.future().cause(), "The write error should be the cause of the failure.");
+        assertNull(onErrorDropped.get(), "No reactive error should be dropped.");
+    }
+
+    /**
+     * Tests that if the promise being used to manage the request-response state succeeds externally while writing the
```
Comment (joshfree) [inline sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java:193]: 'Ensures that the promise managing the request-response state is appropriately handled when it succeeds externally during the request writing process.'

File: sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        assertTrue(promise.future().succeeded(), "The request should have succeeded.");
+        assertEquals(writeError, onErrorDropped.get(), "The write error should be dropped.");
+        Throwable[] writeSuppressed = writeError.getSuppressed();
+        assertEquals(1, writeSuppressed.length, "There should be one suppressed exception.");
+        assertSame(reactiveError, writeSuppressed[0], "The reactive error should be the suppressed exception.");
+    }
+
+    private static void waitForCompletion(Promise<?> promise) throws InterruptedException {
+        Thread.sleep(1000);
+        for (int i = 0; i < 4; i++) {
```
Comment (joshfree) [inline sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java:241]: small nit: make the 4 and 3 literal a single constant used in both places, and just change this loop to <= so you don't need two constants in place of one

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 6]
```diff
@@ -101,6 +102,10 @@ private void write(ByteBuffer bytes) {
                 }
             } else {
                 this.state = State.ERROR;
+                if (error != null) {
```
Comment (joshfree) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:105]: it feels like this check should be a 1-line static helper function we use everywhere

**Review by joshfree — APPROVED** [review]

File: sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java (hunk 5 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/test/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriberTests.java]: This was changed in a commit after this where `i` isn't being checked in the loop anymore.

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 7]
```diff
@@ -122,19 +130,50 @@ private void onErrorInternal(Throwable throwable) {
         // code 2 and greater are completion states which means the error should be dropped as we already completed.
         if (state.code >= 2) {
             Operators.onErrorDropped(throwable, Context.of(contextView));
```
Comment (lmolkova) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:132]: do we need to call it at all from the logic perspective? I'd suggest to log it ourselves and maybe as a warning (unless we know it's severe issue)

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 8]
```diff
@@ -122,19 +130,50 @@ private void onErrorInternal(Throwable throwable) {
...
+
+            // Also, even though Reactor may have an operator for the dropped error or will be logged by Operators
+            // itself, we should log as well as at least this will help associate the error with this class.
+            LOGGER.atError().log(() -> "VertxRequestWriteSubscriber dropped an exception as it already reached a "
```
Comment (lmolkova) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:136]: similar to previous comment: does it need to be an error? Would it result in terminal exception bubbling up to the user code?

Also, I don't think you need lambda here, compiler should optimize string literal concat

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 9]
```diff
@@ -122,19 +130,50 @@ private void onErrorInternal(Throwable throwable) {
...
+        if (!promise.tryFail(throwable)) {
+            // Seems the promise has already completed in some form.
+            // Attempt to associate this error with the existing failure.
+            Throwable cause = promise.future().cause();
+            if (cause != null) {
+                cause.addSuppressed(throwable);
+
+                // Also, even though the exception was added as a suppressed exception to the failed Promise, we should
+                // log as well as at least this will help associate the error with this class.
+                LOGGER.atError().log(() -> "VertxRequestWriteSubscriber added an exception as a suppressed exception "
```
Comment (lmolkova) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:165]: this sounds like an `info` or `warn` if exception is suppressed

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 10]
```diff
@@ -122,19 +130,50 @@ private void onErrorInternal(Throwable throwable) {
...
+                    + "as the Promise already failed.", throwable);
+            } else {
+                // Turns out the future was completed as successfully externally, drop the error.
+                Operators.onErrorDropped(LOGGER.logThrowableAsError(throwable), Context.of(contextView));
```
Comment (lmolkova) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:169]: that combined with the next line will likely result in three errors logged. same as prev comment - do we need to call `onErrorDropped` at all?

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java [hunk 11]
```diff
@@ -122,19 +130,50 @@ private void onErrorInternal(Throwable throwable) {
...
+
+                // Also, even though Reactor may have an operator for the dropped error or will be logged by Operators
+                // itself, we should log as well as at least this will help associate the error with this class.
+                LOGGER.atError().log(() -> "VertxRequestWriteSubscriber dropped an exception as the Promise already "
```
Comment (lmolkova) [inline sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:173]: this sounds like a `debug` log - everything is successful, why log an error?

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java (hunk 6 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:108]: Do you mean across all our Subscriber implementations or this one specifically?

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java (hunk 9 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java]: Changed to info

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java (hunk 10 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:171]: I wanted to respect proper reactive handling

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java (hunk 11 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java]: Changed to info, while the `Promise` succeeded this still failed

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java (hunk 8 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java]: This is the only API that takes Throwable

File: sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java (hunk 7 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/implementation/VertxRequestWriteSubscriber.java:132]: Same question where I wanted to respect reactive flow

**Review by lmolkova — APPROVED** [review]

**Pull request merged**